### PR TITLE
MGMT-11321: Add nutanix support to test infra infrastructure

### PR DESCRIPTION
**Pull request by eliorerz**
- [MGMT-11516](https://issues.redhat.com//browse/MGMT-11516) Add working terraform Nutanix provider resources
- [MGMT-11556](https://issues.redhat.com//browse/MGMT-11556) - Add nutanix platform support.
  - Add NutanixController and NutanixConfig objects.
  - Api/Ingress VIP currently not working as expected.
  - Add nutanix_platform trigger for setting vip_dhcp_allocation.
  - Disable the force flag on `python_terraform` due to a bug.

Note, The test randomly fail to allocate IP to some node/s using terraform, I'm not sure what the issue yet.
```
  :IP address 10.x.x.x is in-use by cloud interface ..
```

/cc @osherdp @vrutkovs @gamli75

### CODE REVIEW

**Comment by eliorerz** [comment]
e2e-metal-assisted-none is expected to fail and will be fixed after https://github.com/openshift/assisted-service/pull/4158

File: src/assisted_test_infra/test_infra/controllers/node_controllers/nutanix_controller.py [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+        vm = self._get_nutanix_vm(tf_vm_name=node_name)
+        vm.power_off()
+        vm.power_on()
+
+    def get_ingress_and_api_vips(self, machine_vips: List[str]):
+        network_subnet_starting_ip = str(
+            ipaddress.ip_address(ipaddress.ip_network(machine_vips[0]).network_address) + 100
+        )
+        ips = utils.create_ip_address_list(2, starting_ip_addr=str(ipaddress.ip_address(network_subnet_starting_ip)))
+        return {"api_vip": ips[0], "ingress_vip": ips[1]}
```
Comment (eliorerz) [inline src/assisted_test_infra/test_infra/controllers/node_controllers/nutanix_controller.py:136]: Need to find a better way to reserve addresses for API/Ingress VIPs

File: src/assisted_test_infra/test_infra/controllers/node_controllers/nutanix_controller.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+from service_client import log
+
+
+class NutanixController(NodeController):
+    _config: BaseNutanixConfig
+
+    def __init__(self, config: BaseNutanixConfig, cluster_config: BaseClusterConfig):
+        super().__init__(config, cluster_config)
+        self.cluster_name = cluster_config.cluster_name.get()
+        folder = TerraformControllerUtil.create_folder(self.cluster_name, platform=cluster_config.platform)
```
Comment (vrutkovs) [inline src/assisted_test_infra/test_infra/controllers/node_controllers/nutanix_controller.py:25]: Does Nutanix have folders?

File: terraform_files/nutanix/variables_nutanix.tf [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+}
+
+
+variable "nutanix_subnet" {
+  type        = string
+  description = ""
+}
+
+
+variable "nutanix_parent_folder" {
```
Comment (vrutkovs) [inline terraform_files/nutanix/variables_nutanix.tf:37]: This seems to be unused

File: terraform_files/nutanix/variables_nutanix.tf [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+  default     = "assisted-test-infra"
+  description = <<EOF
+  Grouping resources together.
+There should be a difference between CI and a manually installed cluster for better management.
+Periodically deleting long-lived CI machines (for example)
+It should be a preexisting folder
+EOF
+}
+
+variable "nutanix_folder" {
```
Comment (vrutkovs) [inline terraform_files/nutanix/variables_nutanix.tf:48]: Also unused

**Review by vrutkovs — CHANGES_REQUESTED** [review]

File: src/assisted_test_infra/test_infra/controllers/node_controllers/nutanix_controller.py (hunk 2 above)
Comment (eliorerz) [reply src/assisted_test_infra/test_infra/controllers/node_controllers/nutanix_controller.py:25]: This is the local terraform folder. I think that Nutanix doesn't have folders.

File: terraform_files/nutanix/variables_nutanix.tf (hunk 3 above)
Comment (eliorerz) [reply terraform_files/nutanix/variables_nutanix.tf:37]: You right, I'll remove it.

Comment (eliorerz) [reply terraform_files/nutanix/variables_nutanix.tf:37]: I thought that we might need it on our CI but now I think that we didn't.
If we will need it in the future I'll re-add it

**Comment by osherdp** [comment]
/uncc
cc me back when it's ready

File: skipper.env [hunk 5]
```diff
@@ -128,4 +128,12 @@ INGRESS_VIP
 WORKERS_COUNT
 MASTERS_COUNT
 IPXE_BOOT
-SUBSYSTEM_LOCAL_REGISTRY
\ No newline at end of file
+SUBSYSTEM_LOCAL_REGISTRY
+NUTANIX_USERNAME
+NUTANIX_PASSWORD
+NUTANIX_ENDPOINT
```
Comment (vrutkovs) [inline skipper.env:134]:
```suggestion
NUTANIX_PRISM_ENDPOINT
```
We should clarify which endpoint is that - Prism Central or Prism Element

File: skipper.env [hunk 6]
```diff
@@ -128,4 +128,12 @@ INGRESS_VIP
...
+NUTANIX_PORT
+NUTANIX_CLUSTER
```
Comment (vrutkovs) [inline skipper.env:136]:
```suggestion
NUTANIX_CLUSTER_NAME
```

File: skipper.env [hunk 7]
```diff
@@ -128,4 +128,12 @@ INGRESS_VIP
...
+NUTANIX_SUBNET
```
Comment (vrutkovs) [inline skipper.env:137]:
```suggestion
NUTANIX_SUBNET_UUID
```

**Review by vrutkovs — COMMENTED** [review]
Looks good overall, lets consider renaming some vars for clarity

File: skipper.env (hunk 7 above)
Comment (eliorerz) [reply skipper.env:137]: This is not subnet uuid, this si the subnet name

File: skipper.env (hunk 5 above)
Comment (eliorerz) [reply skipper.env:134]: This is the same as described in Nutanix terraform documentation https://registry.terraform.io/providers/nutanix/nutanix/latest/docs#endpoint

File: skipper.env (hunk 7 above)
Comment (vrutkovs) [reply skipper.env:137]: Right, lets rename it to SUBNET_NAME, as installer uses UUID for instance

File: src/assisted_test_infra/test_infra/controllers/node_controllers/nutanix_controller.py [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+            vm.update_boot_order([VMBootDevices.DISK, VMBootDevices.CDROM, VMBootDevices.NETWORK])
+
+    def set_per_device_boot_order(self, node_name, key: Callable[[Disk], int]) -> None:
+        raise NotImplementedError
+
+    def get_host_id(self, node_name: str) -> str:
+        raise NotImplementedError
+
+    def set_cpu_cores(self, node_name: str, core_count: int) -> None:
+        raise NotImplementedError
```
Comment (osherdp) [inline src/assisted_test_infra/test_infra/controllers/node_controllers/nutanix_controller.py:194]: can you put all the non-implemented methods at the bottom?

File: src/assisted_test_infra/test_infra/controllers/node_controllers/nutanix_controller.py [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+            self._config.nutanix_password,
+            self._config.nutanix_port,
+            self._config.nutanix_endpoint,
+        )
+
+        clusters = NutanixCluster.list_entities(nutanix_client)
+        try:
+            next(c for c in clusters if c.name == self._config.nutanix_cluster)
+        except StopIteration:
+            raise ValueError(f"Unable to locate nutanix cluster - {self._config.nutanix_cluster}")
```
Comment (osherdp) [inline src/assisted_test_infra/test_infra/controllers/node_controllers/nutanix_controller.py:243]: the more Pythonic way:
```python
for c in NutanixCluster.list_entities(nutanix_client):
    if c.name == self._config.nutanix_cluster:
        break
else:
    raise ValueError("...")
```

**Review by osherdp — CHANGES_REQUESTED** [review]

File: terraform_files/nutanix/variables_nutanix.tf [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+variable "nutanix_port" {
+  type        = number
+  description = "port for the Prism Elements or Prism Central instance. This can also be specified with the NUTANIX_PORT environment variable. Defaults to 9440"
+}
+
+variable "nutanix_cluster" {
+  type        = string
+  description = "Nutanix cluster name"
```
[the remark this answers is not in the document]
Comment (eliorerz) [reply terraform_files/nutanix/variables_nutanix.tf:27]: Not sure actually

File: terraform_files/nutanix/virtual_machines.tf [hunk 11]
```diff
@@ -0,0 +1,77 @@
+
+# Creating the master VMs.
+resource "nutanix_virtual_machine" "master" {
```
[the remark this answers is not in the document]
Comment (eliorerz) [reply terraform_files/nutanix/virtual_machines.tf:3]: This resource declare a single master and not number of masters. On runtime it creates n masters due to the given `masters_count` value.

File: terraform_files/nutanix/virtual_machines.tf [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+    disk_list {
+    disk_size_bytes = var.master_disk
+    device_properties {
+      device_type = "DISK"
+      disk_address = {
+        device_index = 0
+        adapter_type = "SATA"
+      }
+
+    }
```
[the remark this answers is not in the document]
Comment (eliorerz) [reply terraform_files/nutanix/virtual_machines.tf:33]: I don't think there is anything to indent here

File: src/assisted_test_infra/test_infra/tools/terraform_utils.py [hunk 13]
```diff
@@ -70,10 +80,14 @@ def get_state(self) -> Tfstate:
...
     def set_new_vips(self, api_vip: str, ingress_vip: str) -> None:
         self.change_variables(variables={"api_vip": api_vip, "ingress_vip": ingress_vip}, refresh=True)
 
-    def destroy(self) -> None:
-        self.tf.destroy(force=True, input=False, auto_approve=True)
+    def destroy(self, force: bool = True) -> None:
+        if force:
+            self.tf.destroy(force=force, input=False, auto_approve=True)
+        else:
+            self.tf.destroy(input=False, auto_approve=True)
```
[the remark this answers is not in the document]
Comment (eliorerz) [reply src/assisted_test_infra/test_infra/tools/terraform_utils.py:93]: You might be right. I overrided the `destroy` method after writing that conditions.

File: src/assisted_test_infra/test_infra/controllers/node_controllers/nutanix_controller.py [hunk 14]
```diff
@@ -0,0 +1,245 @@
+import ipaddress
+from builtins import list
+from typing import Any, Callable, Dict, List, Tuple, Union
+
+import libvirt
```
[the remark this answers is not in the document]
Comment (eliorerz) [reply src/assisted_test_infra/test_infra/controllers/node_controllers/nutanix_controller.py:5]: Agree

File: src/assisted_test_infra/test_infra/tools/terraform_utils.py [hunk 15]
```diff
@@ -11,12 +11,25 @@
 from service_client import log
 
 
+class _Terraform(Terraform):
+    """python_terraform.Terraform always set the force flag (even to false) causing
+    destroy failures on some cases. This class is overrides the destroy method and set the
+    force flag only if it's set to true"""
+    def destroy(self, dir_or_plan=None, force=False, **kwargs):
```
Comment (vrutkovs) [inline src/assisted_test_infra/test_infra/tools/terraform_utils.py:18]:
```suggestion

    def destroy(self, dir_or_plan=None, force=False, **kwargs):
```
`black` wants a blank line here

Comment (eliorerz) [reply src/assisted_test_infra/test_infra/tools/terraform_utils.py:18]: Yeah, sorry, missed that :|

**Comment by vrutkovs** [comment]
/lgtm
/hold

@osherdp please unhold if you're happy with current state, so that we could proceed with adding a canary CI job

**Comment by eliorerz** [comment]
Locally `e2e-metal-assisted-ipv6` is passing.

/test e2e-metal-assisted-ipv6

File: src/assisted_test_infra/test_infra/tools/terraform_utils.py [hunk 16]
```diff
@@ -11,12 +11,26 @@
 from service_client import log
 
 
+class _Terraform(Terraform):
+    """python_terraform.Terraform always set the force flag (even to false) causing
+    destroy failures on some cases. This class is overrides the destroy method and set the
```
Comment (osherdp) [inline src/assisted_test_infra/test_infra/tools/terraform_utils.py:16]:
```suggestion
    destroy failures on some cases. This class overrides the destroy method and set the
```

**Review by osherdp — CHANGES_REQUESTED** [review]

File: src/assisted_test_infra/test_infra/controllers/node_controllers/nutanix_controller.py [hunk 17]
```diff
@@ -0,0 +1,249 @@
...
+            if vm.power_state != PowerState.OFF.value:
+                log.info(f"Powering off nutanix node {node_name}")
+                vm.power_on()
+            else:
+                log.warning(
+                    f"Attempted to power off node {node_name}, "
+                    f"but the vm is already off - vm.power_state={vm.power_state}"
+                )
+        else:
+            raise ValueError(f"Can't find node with name {node_name}")
```
[the remark this answers is not in the document]
Comment (eliorerz) [reply src/assisted_test_infra/test_infra/controllers/node_controllers/nutanix_controller.py:116]: I remember removing it.. weird..

File: src/assisted_test_infra/test_infra/controllers/node_controllers/nutanix_controller.py [hunk 18]
```diff
@@ -0,0 +1,249 @@
...
+        vm.power_on()
+
+    def get_ingress_and_api_vips(self, is_highly_available: bool = False):
+        nutanix_subnet = next(
+            s for s in NutanixSubnet.list_entities(self._nutanix_client) if s.name == self._config.nutanix_subnet
+        )
+        free_ips = set()
+        ip_network = ipaddress.ip_network(f"{nutanix_subnet.subnet_ip}/{nutanix_subnet.prefix_length}").network_address
+
+        while len(free_ips) < 2:
```
[the remark this answers is not in the document]
Comment (eliorerz) [reply src/assisted_test_infra/test_infra/controllers/node_controllers/nutanix_controller.py:136]: I did that in the first place but I didn't want that if we run 2 tests on the same time we might have a collision

[the remark this answers is not in the document]
Comment (eliorerz) [reply src/assisted_test_infra/test_infra/controllers/node_controllers/nutanix_controller.py:136]: I think it's better to limit the amount of times pinging some random IP

[the remark this answers is not in the document]
Comment (eliorerz) [reply src/assisted_test_infra/test_infra/controllers/node_controllers/nutanix_controller.py:136]: e.g.

```python
        while len(free_ips) < 2:
            address = str(ipaddress.ip_address(ip_network) + random.randint(1, 160))
            if os.system("ping -c 1" + address) > 0:
                free_ips.add(address)

            attempts += 1
            if attempts >= max_ping_attempts:
                raise ConnectionError("Failed to locate free API and Ingress VIPs")
```

[the remark this answers is not in the document]
Comment (osherdp) [reply src/assisted_test_infra/test_infra/controllers/node_controllers/nutanix_controller.py:136]: fine by me, but maybe still scope it with a for loop. e.g.:
```
for _ in range(max_ping_attempts):
    <get_address>
    if <we have 2 addresses>:
        break
else:
    <yell>
```

**Comment by vrutkovs** [comment]
>failed setting platform usages, error is: error adding platform to install config, platform provider wasn't set: no provider with the specified name registered

Seems it breaks e2e-metal-assisted-none, lets try it again to be sure

/test e2e-metal-assisted-kube-api-net-suite

**Comment by osherdp** [comment]
> /test e2e-metal-assisted-none

I suspect it's the change in ``src/assisted_test_infra/test_infra/helper_classes/cluster.py``
```
failed setting platform usages, error is: error adding platform to install config, platform provider wasn't set: no provider with the specified name registered
```
I wonder why the assisted-service logs don't just say which value it got that matches no existing provider

**Comment by vrutkovs** [comment]
Seems we need https://github.com/openshift/assisted-service/pull/4158, but according to `skopeo inspect -n docker://registry.ci.openshift.org/edge-infrastructure/ocm-2.7@sha256:8e449c94f280d44cf3b32aaa9601a01accce397e8c387c08cd042c974a7ab5af` we had just https://github.com/openshift/assisted-service/commits/1f13a9b65a96d79ff5ce82b41610cb3ac972c272

**Comment by osherdp** [comment]
ack. let's wait for image promotion to finish

**Comment by vrutkovs** [comment]
assisted-kube-api-net-suite seems to be failing in other PRs too

**Comment by eliorerz** [comment]
> e2e-metal-assisted-none is expected to fail and will be fixed after [openshift/assisted-service#4158](https://github.com/openshift/assisted-service/pull/4158)

@vrutkovs None platform test fixed as part of 4150

**Comment by eliorerz** [comment]
@osherdp @vrutkovs Please ensure that all comments were fixed.
Thanks

**Comment by vrutkovs** [comment]
/lgtm
Leaving hold for Osher

**Review by osherdp — APPROVED** [review]

**Pull request merged**